java/arcs/core/analysis/ClaimDeduction.kt

```diff
@@ -0,0 +1,104 @@
+
+internal typealias Path = List<String>
+internal typealias Paths = List<Path>
+internal typealias ClaimDerivations = Map<Path, Set<Path>>
+
+fun <E> MutableList<E>.push(element: E) = add(element)
+fun <E> MutableList<E>.pop(): E = removeAt(size - 1)
+fun <E> MutableList<E>.peek(): E = this[count()]
+fun <E> MutableList<E>.popOrDefault(default: E): E = if (size == 0) default else pop()
+fun <E> MutableList<E>.peekOrDefault(default: E): E = if (size == 0) default else peek()
+
+/** A visitor that accumulates [ClaimDerivations] from a Paxel [Expression]. */
+class ClaimDeducer : Expression.Visitor<ClaimDerivations> {
+
+    /** Stack of intermediate [Path]s; they have no derivation (yet). */
+    /* internal */ val stack = mutableListOf<Paths>()
+
+    override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun <L, R, T> visit(expr: Expression.BinaryExpression<L, R, T>): ClaimDerivations {
+        val result = expr.left.accept(this) + expr.right.accept(this)
+
+        stack.push(stack.pop() + stack.pop())
+
+        return result
+    }
+
+    override fun <T> visit(expr: Expression.FieldExpression<T>): ClaimDerivations {
+        // Base case: unqualified field must refer to handle connection.
+        if (expr.qualifier == null) {
+            stack.push(listOf(listOf(expr.field)))
+            return emptyMap()
+        }
+
+        // Recursive case: update path on the stack with a child field.
+        val previous = expr.qualifier!!.accept(this)
+        stack.push(listOf(stack.popOrDefault(listOf(listOf()))[0] + listOf(expr.field)))
+
+        return previous
+    }
+
+    override fun <T> visit(expr: Expression.QueryParameterExpression<T>): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.NumberLiteralExpression): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.TextLiteralExpression): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.BooleanLiteralExpression): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.FromExpression): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.WhereExpression): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun <T> visit(expr: Expression.SelectExpression<T>): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.LetExpression): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun <T> visit(expr: Expression.FunctionExpression<T>): ClaimDerivations {
+        TODO("Not yet implemented")
+    }
+
+    override fun visit(expr: Expression.NewExpression) =
+        expr.fields.fold(emptyMap<Path, Set<Path>>()) { acc, (key, rhs) ->
```

Comment:
This will be cleaner and easy to follow with [associateBy](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/associate-by.html#associateby).